### PR TITLE
sql: allow adding a primary key when table has default rowid primary key

### LINKED ISSUE #45509: sql: make allow for `add constraint "primary"` when a table has the default rowid primary key
Consider the following table:

```
CREATE TABLE t (x INT);
```

I should be able to run `ALTER TABLE t ADD CONSTRAINT "primary" ...` to give this table a primary key and remove the old rowid column.

### PR DESCRIPTION
**Pull request by rohany**
Fixes #45509.

Release note (sql change): This commit allows for users to use the
`ALTER TABLE .. ADD PRIMARY KEY ...` command when the target table
has the default rowid primary key.

### CODE REVIEW

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45514)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, and @rohany)

___
*[pkg/sql/alter_table.go, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45514#-M17oeb842NfhuTnFMo_:-M17oeb842NfhuTnFMoa:bgufjny) ([raw file](https://github.com/cockroachdb/cockroach/blob/be2a81df8998e2c82fd2a93ba5687cbac09a2a4e/pkg/sql/alter_table.go#L112)):*
> ```Go
> 
> 	for i := 0; i < len(n.n.Cmds); i++ {
> 		cmd := n.n.Cmds[i]
> ```

Why this change? It doesn't do anything memory related unless you say `cmd := &n.n.Cmds[i]` for example.

Oh I see, it's because you're editing the loop... you need to add a comment about that I think.
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 603 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45514#-M17to5Y1XFsSC56uZ02:-M17to5Y1XFsSC56uZ03:b-doxg5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/be2a81df8998e2c82fd2a93ba5687cbac09a2a4e/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L603)):*
> ```
> 
> statement ok
> ALTER TABLE t ADD PRIMARY KEY (x)
> ```

oh yeah, now we're talking my dude
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 615 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45514#-M17u7M81zcwtlygDakc:-M17u7M81zcwtlygDakd:b-d7cmqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/be2a81df8998e2c82fd2a93ba5687cbac09a2a4e/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L615)):*
> ```
> 
> statement ok
> DROP TABLE IF EXISTS t1, t2 CASCADE;
> ```

nit: IF EXISTS isn't necessary...
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 618 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45514#-M17uNpF2iyjt8GBPs6N:-M17uNpF2iyjt8GBPs6O:b-s0027x) ([raw file](https://github.com/cockroachdb/cockroach/blob/be2a81df8998e2c82fd2a93ba5687cbac09a2a4e/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L618)):*
> ```
> CREATE TABLE t1 (x INT PRIMARY KEY);
> CREATE TABLE t2 (x INT NOT NULL, y INT NOT NULL, FAMILY (x, y));
> ALTER TABLE t2 ADD PRIMARY KEY (x, y) INTERLEAVE IN PARENT t1 (x)
> ```

whoa, this works? so you're saying we can edit tables to be interleaved now? this is huge...

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45514)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, and @rohany)

___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 615 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45514#-M17u7M81zcwtlygDakc:-M17uxgJEVmZX-KQkQub:bqh6irf) ([raw file](https://github.com/cockroachdb/cockroach/blob/be2a81df8998e2c82fd2a93ba5687cbac09a2a4e/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L615)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: IF EXISTS isn't necessary...
</blockquote></details>

there are subtests littered throughout this file -- we need the if exists so that running a subtest doesn't error.
___
*[pkg/sql/logictest/testdata/logic_test/alter_primary_key, line 618 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45514#-M17uNpF2iyjt8GBPs6N:-M17uwP66vAVRy4Wxfbr:b-jp3km) ([raw file](https://github.com/cockroachdb/cockroach/blob/be2a81df8998e2c82fd2a93ba5687cbac09a2a4e/pkg/sql/logictest/testdata/logic_test/alter_primary_key#L618)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

whoa, this works? so you're saying we can edit tables to be interleaved now? this is huge...
</blockquote></details>

The future is now old man

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45514)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @lucy-zhang, and @rohany)

___
*[pkg/sql/alter_table.go, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45514#-M17oeb842NfhuTnFMo_:-M183AMk-gkAtpAUy4bh:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/be2a81df8998e2c82fd2a93ba5687cbac09a2a4e/pkg/sql/alter_table.go#L112)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why this change? It doesn't do anything memory related unless you say `cmd := &n.n.Cmds[i]` for example.

Oh I see, it's because you're editing the loop... you need to add a comment about that I think.
</blockquote></details>

done

File: pkg/sql/parser/sql.y [hunk 1]
```diff
@@ -4737,12 +4737,13 @@ constraint_elem:
       },
     }
   }
-| PRIMARY KEY '(' index_params ')' opt_hash_sharded
+| PRIMARY KEY '(' index_params ')' opt_interleave opt_hash_sharded
```
Comment (jordanlewis) [inline pkg/sql/parser/sql.y]: I forgot to mention but you should have a parse test for this. Other than this lgtm

**Comment by rohany** [comment]
updated with some parse tests.

**Review by jordanlewis — APPROVED** [review]

**Comment by rohany** [comment]
bors r=jordanlewis

**Pull request merged**